### PR TITLE
acl filters are applied based on user input not saved object

### PR DESCRIPTION
**Pull request by soldair**
where acl filters change based on an object property
the acl was applied in the context of the users changes rather than the
saved object. This bug allowed users to ignore filter settings on updates.

### CODE REVIEW

**Comment by soldair** [comment]
i dont know why the build failed on travis. It passes for when i run them. ill take a look but considering it failed on my last pr that only added a debug flag it seems to be some travs env thing.

**Comment by soldair** [comment]
@rjrodger @nherment review/merge?

File: lib/ACLMicroservicesBuilder.js [hunk 1]
```diff
@@ -229,42 +226,39 @@ ACLMicroservicesBuilder.prototype._executeSavePermissions = function(args, callb
 
   if(args.ent.id) { // update
 
-    self.ACLMicroservicesBuilder._loadAndAuthorize(entityDef, args.ent.id, args.cmd, roles, context, args.showSoftDenied$, function(err, dbEntity) {
+    self.ACLMicroservicesBuilder._loadAndAuthorize(entityDef, args.ent.id, args.cmd, roles, context, args.showSoftDenied$,args.ent, function(err, dbEntity) {
```
Comment (nherment) [inline lib/ACLMicroservicesBuilder.js]: not sure why you need to pass in the entity there. 
My understanding is that the bug is in the save_existing and the effect is that filters do not get run on save.
``_loadAndAuthorize`` is supposed to check if you have read access to that entity and executes all the 'load' ACLs. It should not need to know about the entity being saved. That is done further below.

Comment (soldair) [inline lib/ACLMicroservicesBuilder.js]: args.ent gets filtered based on the filter rules from the state of dbEntity which is loaded inside _loadAndAuthorize.

**Comment by soldair** [comment]
### best explanation so far

in every case the conditions comparison must be done on the last version of the object. If this comparison provides filters that prevent fields from being changed they need to be enforced on the changes to the object rather than being ignored.

File: lib/ACLMicroservicesBuilder.js [hunk 2]
```diff
@@ -229,42 +226,39 @@ ACLMicroservicesBuilder.prototype._executeSavePermissions = function(args, callb
...
-
-          if(err) {
+        return callback(err, undefined)
 
-            callback(err, undefined)
+      }
 
-          } else {
+      // also execute permission checks on the new attributes
+      self.ACLMicroservicesBuilder._deepAuthorize(entityDef, args.ent, args.cmd, 'save_existing', roles, context, true, args.showSoftDenied$, function(err, filteredEntity) {
```
Comment (nherment) [inline lib/ACLMicroservicesBuilder.js]: it looks that a ``save_new`` changed to a ``save_existing``. It changes the semantics of these 2 actions.

Comment (soldair) [inline lib/ACLMicroservicesBuilder.js]: it doesn't make a difference still. they all become "save" by the time they are in perms land i think. but ill confirm that this is what i want

**Comment by nherment** [comment]
sorry I'm having troubled understanding the change and its consequences.
If all our client's tests pass (including the 2 nearform projects for your current client) then you have the freedom to make any change but I'd prefer you to make it on a 0.6.0 version so that we can still fallback on 0.5.X if we need to.

**Comment by soldair** [comment]
ok sounds good.
I would really be more comfortable if i could explain it =). That's why the bug exists to begin with.

ill describe it in steps. but the test shows it's consequences. and this test fails on the older version.
- i have an object in the database. Only admins can change a case sytatus once it's closed, but other properties can be changed.
```
{id:1,status:'closed',name:'bob'}
```

```
...conditions:[{attributes:{status:"closed"}}],filters:{closed:false}...
```
- i create an update that changes status without permission

```
roles = []

e = ....load$(id)
e.status = "active"
e.name = "changed bob"
e.save$(..
 console.log(e)
```
before the change this code would log something like

```
{status:"active",name:"changed bob"}
```
which is bad because it's explicitly not allowed to change status! 

after the change
```
{status:"closed",name:"changed bob"}
```
note because only changes to status were filtered only changes to status were prevented.

in isight only some people are allowed to re open cases but it intersects with people who can update properties of closed cases. enforcing this constraint is not possible with this bug.

**Comment by nherment** [comment]
@rjrodger can you please give @soldair rights on this repo ?
I already took care of his npm access.

**Comment by soldair** [comment]
if you think its good now you could merge. otherwise @rjrodger can review too and we can get this version published.

**Pull request merged**